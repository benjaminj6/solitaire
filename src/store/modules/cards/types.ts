export type CardRank =
  | 'A'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | 'J'
  | 'Q'
  | 'K';

export type CardRankNumber =
  | '0'
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | '11'
  | '12';

export type Suit = 'hearts' | 'spades' | 'diamonds' | 'clubs';

// TODO - make the typing on this more strict to basically just be a suit and a rank separated by a dash
export type CardId = string;

export type Card = {
  id: CardId;
  suit: Suit;
  rank: CardRank;
  value: number;
  visible: boolean;
};

export type Deck = {
  [x: string]: Card;
};

export type CardRefArray = CardId[];

export type TableauRow = '1' | '2' | '3' | '4' | '5' | '6' | '7';
export type Tableau = {
  [row in TableauRow]: CardRefArray;
};

export type Stock = CardRefArray;
export type Hand = CardRefArray;

export type Foundation = { [suit in Suit]: CardRefArray };

export type State = {
  deck: Deck;
  dealt: boolean,
  tableau: Tableau;
  stock: Stock;
  hand: Hand;
  foundation: Foundation;
};
