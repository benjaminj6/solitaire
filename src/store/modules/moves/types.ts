import { CardId, CardLocation } from '../cards/types';

export type Move = {
  cards: CardId[];
  currentLocation?: CardLocation;
  destination?: CardLocation;
};

export type MoveId = string;

export type State = {
  nextMove: Move | {};
  moves: {
    [id: string]: Move;
  };
  history: Move[];
};
