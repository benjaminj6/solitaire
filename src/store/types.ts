// Generic Redux-related types
import { Store, Dispatch } from 'redux';

export type Action = {
  type: string;
  result?: {
    [x: string]: any;
  };
};

export type ActionCreator = (...args: any[]) => Action;

type GetState = Store<any>['getState'];
export type ActionThunk = (dispatch: Dispatch<Action>, getState: GetState) => Action;

export type ActionThunkCreator = (...args: any[]) => ActionThunk;

export type Reducer < T > = (state: T, action: Action) => T;
