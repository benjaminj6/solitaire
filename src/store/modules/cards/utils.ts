import { Tableau, TableauRow, CardRefArray } from './types';

export const shuffleArray = (originalArray: any[]) => {
  // a little higher on memory, but small cost to pay for guaranteed immutability
  let arr = [...originalArray];
  let current = arr.length;
  let temp;
  let i;

  while (current) {
    i = Math.floor(Math.random() * current--)

    temp = arr[current]
    arr[current] = arr[i]
    arr[i] = temp
  }

  return arr;
}

const getNonFilledRow = (tableau: Tableau): string | null => {
  const keys = Object.keys(tableau).sort((a, b) => +a - +b);
  const rowNumber = keys.find((key: TableauRow) => tableau[key].length < +key);
  return rowNumber ? `${rowNumber}` : null;
};

type DealTableauCards = (cards: CardRefArray, tableau: Tableau) => Tableau;
export const dealTableauCards: DealTableauCards = (cards, originalTableau) => {
  const { tableau: dealtTableau } = cards.reduce(
    (acc: { row: TableauRow; tableau: Tableau }, card, i) => {
      const { row, tableau } = acc;

      if (row) {
        tableau[row].push(card);
      }

      return {
        row:
          row === '7' || row === null
            ? getNonFilledRow(acc.tableau)
            : `${+row + 1}`,
        tableau: acc.tableau,
      };
    },
    {
      row: getNonFilledRow(originalTableau),
      tableau: originalTableau,
    },
  );

  return dealtTableau;
};
