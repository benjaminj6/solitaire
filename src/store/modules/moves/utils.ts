import {
  CardId,
  CardLocation,
  Deck,
  Foundation,
  FoundationRowLocation,
  Tableau,
} from '../cards/types';
import { validateCardBase, validateFoundationMove } from '../cards/utils';
import { first, get } from 'lodash';

// TODO -- perhaps this should be moved into the central `cards` reducer, & the side-effect is the logging of actions into history
type ValidateMove = (
  args: {
  cards: CardId[];
  nextLocation: CardLocation;
  deck: Deck;
  foundation: Foundation;
  tableau: Tableau;
  }
) => boolean;
export const validateMove: ValidateMove = ({
  cards,
  nextLocation,
  deck,
  foundation,
  tableau,
}) => {
  const isValidTableauMove =
    nextLocation.includes('tableau') &&
    validateCardBase(deck[first(cards)], deck[get({ tableau }, nextLocation, {})]);
  const isValidFoundationMove =
    nextLocation.includes('foundation') &&
    cards.length === 1 &&
    validateFoundationMove({
      destination: nextLocation as FoundationRowLocation,
      foundationCards: get({ foundation }, nextLocation, []) as CardId[],
      card: deck[first(cards)],
      deck,
    });

  return isValidFoundationMove || isValidTableauMove;
};
