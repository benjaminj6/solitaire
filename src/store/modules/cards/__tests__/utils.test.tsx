import { dealTableauCards, shuffleArray } from '../utils';
import { range } from 'lodash';
import { Tableau } from '../types'

describe('shuffleArray', () => {
  const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  test('should return an array', () => {
    // note - since this function uses Math.random under the hood,
    // I wonder if there's a chance that it could break. Look into what the
    // testing process is for things like this
    expect(shuffleArray(arr)).not.toEqual(arr);
  });
});

describe('dealTableauCards', () => {
  const cards = range(0, 25).map(num => `${num}`);
  const tableau: Tableau = {
    '1': [],
    '2': [],
    '3': [],
    '4': [],
    '5': [],
    '6': [],
    '7': [],
  };

  test('should disperse the cards to the tableau rows', () => {
    const result = dealTableauCards(cards, tableau);
    expect(result).toMatchSnapshot();
  });
});