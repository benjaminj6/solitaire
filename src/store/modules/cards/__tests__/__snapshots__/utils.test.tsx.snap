// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dealTableauCards should disperse the cards to the tableau rows 1`] = `
Object {
  "1": Array [
    "0",
  ],
  "2": Array [
    "1",
    "7",
  ],
  "3": Array [
    "2",
    "8",
    "13",
  ],
  "4": Array [
    "3",
    "9",
    "14",
    "18",
  ],
  "5": Array [
    "4",
    "10",
    "15",
    "19",
    "22",
  ],
  "6": Array [
    "5",
    "11",
    "16",
    "20",
    "23",
  ],
  "7": Array [
    "6",
    "12",
    "17",
    "21",
    "24",
  ],
}
`;
